### PR TITLE
[RSDK-1155] upgrade deps

### PR DESCRIPTION
**Pull request by nicksanford**
Can only be reviewed once https://github.com/erh/viamrtsp/pull/16 is merged

Changes:
1. Update RDK to https://github.com/viamrobotics/rdk/releases/tag/v0.25.0-rc2
1. change `cmd/module/cmd.go` and `cmd/remote/cmd.go`  to use `utils.ContextualMain` so RTSP connections get closed when the module process  is sent kill / quit signals
1. Upgrade gortsplib from `v3.7.0` to `v4.8.0` (this was a pre req for getting rtp_passthrough to work)
1. Upgrade goutils from `v0.1.60` to `v0.1.71`
1. Upgrade rtp from `1.8.2` to `1.8.5`
1. Upgrade from go `1.19` to `1.21`
2. Add `github.com/bluenviron/mediacommon v1.9.2` (needed due to `gortsplib` upgrade
3. Upgrade `github.com/pion/webrtc/v3` to  `v3.2.36` from `v3.2.21`
4. Add `viamrtsp.SetLibAVLogLevelFatal()` function to disable all non fatal libav logs. Only called during tests
1. Add tests for GetImage
2. Change condition for reconnection based on `OPTIONS` response & error

Manual Testing: (with vaim-server https://github.com/viamrobotics/rdk/releases/tag/v0.24.1)
- Change rtsp camera from outputting h264 -> h265 -> h264B -> H264H & confirm that the camera stream reconnects (does not apply for `erh:viamrtsp:rtsp-mjpeg`)
- for non  `erh:viamrtsp:rtsp` confirm that camera streams with the codec in their model name work & all others prevent them from booting
- trigger reconfiguration 

Models:
- [x] `erh:viamrtsp:rtsp`
- [x] `erh:viamrtsp:rtsp-h264`
- [x] `erh:viamrtsp:rtsp-h265`
- [x] `erh:viamrtsp:rtsp-mjpeg`

Platforms:
- [x] Macos aarm64 14.4.1
- [x] Ubuntu 22.04.4 amd64
- [x] RPI 4 (bullseye) arm64
- [x] RPI 5 (bookworm) arm64

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,13 +1,19 @@
 module github.com/erh/viamrtsp
 
-go 1.19
+go 1.21
+
+toolchain go1.21.8
```
Comment (seanavery) [inline go.mod:5]: [info] https://go.dev/doc/toolchain

had not seen this before! is there a reason why we are setting toolchain to 1.21.8?

File: rtsp.go [hunk 2]
```diff
@@ -114,53 +116,61 @@ func (rc *rtspCamera) clientReconnectBackgroundWorker(codecInfo videoCodec) {
...
+
+	rc.closeConnection()
 
 	// replace the client with a new one, but close it if setup is not successful
 	rc.client = &gortsplib.Client{}
+	rc.client.OnPacketLost = func(err error) {
+		rc.logger.Debugf("OnPacketLost: err: %s", err.Error())
+	}
+	rc.client.OnTransportSwitch = func(err error) {
+		rc.logger.Debugf("OnTransportSwitch: err: %s", err.Error())
```
Comment (seanavery) [inline rtsp.go:165]: Have we been able to test transport switches? Should we fire off a reconnect in this situation?

File: rtsp_test.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+	"go.viam.com/test"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/rimage/transform"
+)
+
+func TestRTSPCamera(t *testing.T) {
```
Comment (seanavery) [inline rtsp_test.go:27]: Should we be running this as apart of the CI flow?

File: rtsp.go [hunk 4]
```diff
@@ -110,57 +112,67 @@ func (rc *rtspCamera) clientReconnectBackgroundWorker(codecInfo videoCodec) {
...
-					errors.Is(err, syscall.EPIPE) ||
-					errors.Is(err, syscall.ECONNREFUSED)) {
-					rc.logger.Warnw("The rtsp client encountered an error, trying to reconnect", "url", rc.u, "error", err)
+				rc.logger.Debugf("Options reponse: %s, err: %s", res, err)
+				// Nick S:
+				// This error happens all the time on hardware we need to support & does not affect
+				// the performance of camera streaming. As a result, we ignore this error specifically
+				_, isErrClientInvalidState := err.(liberrors.ErrClientInvalidState)
+				if err != nil && !isErrClientInvalidState {
+					rc.logger.Warnf("The rtsp client encountered an error, trying to reconnect to %s, err: %s", rc.u, err)
```
Comment (nicksanford) [inline rtsp.go:121]: After upgrading to gortsplib v4 we were occasionally seeing `syscall.ECONNRESET` errors which did not trigger reconnection but did cause result in the camera no longer receiving RTP packets, which caused the connection to go dead & the camera to freeze.

As a result I changed this logic to consider all errors other than the specific error we see all the time as a condition that should trigger reconnection.

File: go.mod (hunk 1 above)
Comment (nicksanford) [reply go.mod:5]: it was automatically set by go 1.21 when I ran `go mod tidy`

File: rtsp.go (hunk 2 above)
Comment (nicksanford) [reply rtsp.go:167]: I have not seen this triggered no

File: rtsp_test.go (hunk 3 above)
Comment (nicksanford) [reply rtsp_test.go:27]: Yes, I was going to add a ticket to do so after this is merged

File: rtsp.go [hunk 5]
```diff
@@ -213,141 +226,141 @@ func (rc *rtspCamera) reconnectClient(codecInfo videoCodec) (err error) {
...
 	if err != nil {
-		return err
+		return errors.Wrap(err, "creating H265 raw decoder")
 	}
 
 	// For H.265, handle VPS, SPS, and PPS
-	if format.VPS != nil {
-		rc.rawDecoder.decode(format.VPS)
+	if f.VPS != nil {
+		rc.rawDecoder.decode(f.VPS) // nolint:errcheck
```
Comment (randhid) [inline rtsp.go:327]: Same as in your other pr, were we just not lining and never checking the error? Okay to address in the other pr.

**Review by randhid — APPROVED** [review]

File: rtsp.go (hunk 5 above)
Comment (nicksanford) [reply rtsp.go:325]: Yes. We are not running lint in CI (we should change that) and we were not checking the error

File: rtsp_test.go (hunk 3 above)
Comment (nicksanford) [reply rtsp_test.go:27]: https://viam.atlassian.net/browse/RSDK-7316

File: rtsp.go (hunk 5 above)
Comment (nicksanford) [reply rtsp.go:325]: https://viam.atlassian.net/browse/RSDK-7315

**Review by seanavery — APPROVED** [review]

**Pull request merged**